File: fx/src/main/java/gargoyle/fx/FXWrapper.java
package gargoyle.fx;

import gargoyle.fx.log.FXLog;
import javafx.application.Application;
import javafx.stage.Stage;

import java.awt.Component;
import java.awt.Window;
import java.util.Arrays;
import java.util.Map;

public final class FXWrapper extends Application {
    private static final String LC_TITLE = "title";
    private Class<? extends FXApplication> appClass;
    private String appClassName;
    private FXContext context;

    public FXWrapper() {
    }

    @SuppressWarnings("unchecked")
    @Override
    public void init() {
        try {
            appClassName = getParameters().getRaw().get(0);
            appClass = FXIntUtil.loadClass(appClassName, getClass().getClassLoader());
            context = FXContextFactory.currentContext();
            context.register(appClassName, (Class<FXApplication>) appClass, (Class<FXApplication>) appClass);
            context.initializeContext(getHostServices(), getParameters());
            FXLauncher.notifySplash(FXSplash.FXSplashNotification.Type.INIT, 0.0, "");
            final FXApplication application = context.getBean(appClassName, appClass);
            application.doInit();
        } catch (final Exception e) {
            FXLauncher.splashStop("");
            FXLog.error(e, "{appClassName} initialization error", Map.of("appClassName", appClassName));
            throw new FXUserException(e, "{appClassName} initialization error", Map.of("appClassName", appClassName));
        }
    }

    @Override
    public void start(final Stage primaryStage) {
        try {
            context.register(Stage.class, primaryStage);
            primaryStage.getProperties().put(FXLauncher.KEY_APP, this);
            primaryStage.setOnCloseRequest(event -> {
                try {
                    if (Arrays.stream(FXCloseFlag.values()).anyMatch(closeFlag -> FXCloseAction.KEEP == closeFlag.doIf(context))) {
                        event.consume();
                    }
                } catch (final Exception e) {
                    FXLog.error(e, e.getLocalizedMessage());
                }
            });
            FXLauncher.notifySplash(FXSplash.FXSplashNotification.Type.START, 0.0, "");
            FXIntUtil.prepareStage(context, FXUtil.resolveBaseName(appClass), primaryStage);
            FXUtil.loadResources(context.getBaseClass(), context.getLocale(),
                    FXUtil.resolveBaseName(appClass)).ifPresent(resources ->
                    primaryStage.setTitle(resources.getString(LC_TITLE)));
            primaryStage.setScene(FXUtil.getOrCreateScene(context.getBean(appClassName, appClass).doStart().getView()));
            primaryStage.centerOnScreen();
            primaryStage.show();
            FXLauncher.splashStop("");
        } catch (final Exception e) {
            FXLauncher.splashStop(FXUtil.format(
                    "{appClassName} initialization error", Map.of("appClassName", appClassName)));
            FXLog.error(e, "{appClassName} initialization error", Map.of("appClassName", appClassName));
            throw new FXUserException(e, "{appClassName} initialization error", Map.of("appClassName", appClassName));
        }
    }

    @Override
    public void stop() {
        try {
            askStop();
            Arrays.stream(Window.getWindows()).filter(Component::isDisplayable).forEach(Window::dispose);
        } catch (final Exception e) {
            FXLog.error(e, "{appClassName} initialization error", Map.of("appClassName", appClassName));
            throw new FXUserException(e, "{appClassName} initialization error", Map.of("appClassName", appClassName));
        }
    }

    void askStop() throws Exception {
        try {
            context.getBean(appClassName, appClass).doStop();
        } finally {
            FXLauncher.splashStop(FXUtil.format("{appClassName} initialization error",
                    Map.of("appClassName", appClassName)));
            context.getRegistry().close();
        }
    }

}

File: fx/src/main/kotlin/gargoyle/fx/FxWrapper.kt
package gargoyle.fx

import gargoyle.fx.FxSplash.FxSplashNotification
import gargoyle.fx.FxUtil.get
import gargoyle.fx.log.FxLog
import javafx.application.Application
import javafx.event.EventHandler
import javafx.stage.Stage
import java.awt.Window
import kotlin.reflect.KClass

class FxWrapper : Application() {
    private lateinit var appClass: KClass<out FxApplication>
    private lateinit var appClassName: String
    private lateinit var context: FxContext

    @Suppress("UNCHECKED_CAST")
    override fun init(): Unit =
        try {
            appClassName = parameters.raw[0]
            appClass = FxIntUtil.loadClass(appClassName, javaClass.classLoader)
            context = FxContext.current
            context.register(appClassName, FxApplication::class, appClass)
            context.initializeContext(hostServices, parameters)
            FxLauncher.notifySplash(FxSplashNotification.Type.INIT, 0.0, "")
            context.getBean(appClassName, FxApplication::class).doInit()
        } catch (e: Exception) {
            FxLauncher.splashStop("")
            FxLog.error("$appClassName initialization error", e)
            throw FxException("$appClassName initialization error", e)
        }

    override fun start(primaryStage: Stage): Unit =
        try {
            context.register(Stage::class, primaryStage)
            primaryStage.properties[FxLauncher.KEY_APP] = this
            with(primaryStage) {
                onCloseRequest = EventHandler { event ->
                    try {
                        if (FxCloseFlag.values().any { FxCloseAction.KEEP == it.doIf(context) }) event.consume()
                    } catch (e: Exception) {
                        FxLog.error(e.localizedMessage, e)
                    }
                }
                FxLauncher.notifySplash(FxSplashNotification.Type.START, 0.0, "")
                FxIntUtil.prepareStage(FxUtil.resolveBaseName(appClass), this, context)
                FxUtil.loadResources(context.baseClass, context.locale, FxUtil.resolveBaseName(appClass))
                    ?.let { title = it[LC_TITLE] }
                scene = FxUtil.getOrCreateScene(context.getBean(appClassName, FxApplication::class).doStart().view)
                centerOnScreen()
                show()
            }
            FxLauncher.splashStop("")
        } catch (e: Exception) {
            FxLauncher.splashStop("$appClassName initialization error")
            throw FxException("$appClassName initialization error", e)
        }

    override fun stop() {
        try {
            askStop()
            Window.getWindows().filter { it.isDisplayable }.forEach { it.dispose() }
        } catch (e: Exception) {
            throw FxException("$appClassName initialization error", e)
        }
    }

    fun askStop() {
        try {
            context.getBean(appClassName, FxApplication::class).doStop()
        } finally {
            FxLauncher.splashStop("$appClassName initialization error")
            context.registry.close()
        }
    }

    companion object {
        private const val LC_TITLE = "title"
    }
}
